### PR TITLE
Reduce memory usage in process_centr_segment

### PR DESCRIPTION
**Pull request by LachlanStuart**
See individual commit messages.

I still need to check that the results are still valid. They're pretty close, but I have to check why a few discrepancies are reported.

### CODE REVIEW

File: annotation_pipeline/validate.py [hunk 1]
```diff
@@ -104,14 +108,18 @@ def formula_image_metrics(formula_images_it, compute_metrics):
...
             if f_metrics['msm'] > 0:
                 formula_metrics[f_i] = f_metrics
                 formula_images[f_i] = f_images
 
     for f_i, p_i, f_int, image in formula_images_it:
-        formula_images_buffer[f_i][p_i] = image
+        if formula_images_buffer[f_i][p_i] is None:
+            formula_images_buffer[f_i][p_i] = image
+        else:
+            formula_images_buffer[f_i][p_i] += image
```
Comment (omerb01) [inline annotation_pipeline/validate.py:122]: I suspect that this part causes incorrect metrics

**Review by omerb01 — CHANGES_REQUESTED** [review]
@LachlanStuart I think it is necessary to load database segments lazily because the reason you explained in Slack - we don't want a scenario that loads too many dataset segments into one action.
I tested your PR with "big" dataset and I got wrong output:
```
2019-12-11 18:50:51,942 [INFO] annotation-pipeline: Missing annotations: 0 (PASS)
2019-12-11 18:50:51,943 [ERROR] annotation-pipeline: Incorrect spatial metric: 8 (FAIL)
2019-12-11 18:50:51,944 [ERROR] annotation-pipeline: Incorrect spectral metric: 22 (FAIL)
2019-12-11 18:50:51,944 [INFO] annotation-pipeline: Incorrect chaos metric: 0 (PASS)
2019-12-11 18:50:51,945 [ERROR] annotation-pipeline: Incorrect MSM: 10 (FAIL)
2019-12-11 18:50:51,946 [INFO] annotation-pipeline: FDR changed: 57 (PASS)
2019-12-11 18:50:51,946 [INFO] annotation-pipeline: FDR changed significantly: 0 (PASS)
2019-12-11 18:50:51,951 [ERROR] annotation-pipeline: Incorrect spatial metric extra info:
          spatial  spatial_ref     error
ion_i                                   
2818099  0.458322     0.421547  0.036775
1606071  0.450914     0.444191  0.006723
4464929  0.079508     0.085761  0.006253
4156134  0.059229     0.054041  0.005188
3083069  0.060633     0.065090  0.004457

2019-12-11 18:50:51,955 [ERROR] annotation-pipeline: Incorrect spectral metric extra info:
         spectral  spectral_ref     error
ion_i                                    
4464929  0.906230      0.450100  0.456130
4156134  0.900490      0.523679  0.376811
2818099  0.985673      0.708021  0.277652
2342387  0.736505      0.578693  0.157812
373001   0.494198      0.468673  0.025525

2019-12-11 18:50:51,958 [ERROR] annotation-pipeline: Incorrect MSM extra info:
              msm   msm_ref     error
ion_i                                
2818099  0.451440  0.298256  0.153184
4464929  0.071496  0.038303  0.033193
4156134  0.053229  0.028244  0.024985
1606071  0.408331  0.401638  0.006693
2342387  0.012388  0.009894  0.002494

2019-12-11 18:50:51,958 [ERROR] annotation-pipeline: 3 checks failed
```
and also I found that 60746 metrics calculated instead of 60745.

**Comment by gilv** [comment]
@omerb01 we still need this one?

**Comment by omerb01** [comment]
@gilv yes, to get more precise control on `annotate` method's memory

**Comment by LachlanStuart** [comment]
@omerb01 I believe the changes in this PR aren't compatible with the changes from your PR. As I described in Slack, this PR requires `centr_df` to be iterated in `mz` order so that dataset segments can be lazily loaded, and your PR changed it to be iterated in `formula_i` order so that images are produced in full sets, so that they don't need to be kept in memory for very long.

If you feel there's value in keeping this implementation, I'll make an alternate function for it and add a keyword param for switching between the two strategies. However, I believe that the worst-case memory usage of your implementation should still fit in 2GB of memory for the 60GB dataset, so I'm not sure if it's necessary...

**Comment by omerb01** [comment]
@LachlanStuart I would like that any dataset in any size won't cause OOM in future.
do you think it isn't possible to iterate over dataset segments and also to get all images related for a specific formula each time?
I sorted the database by `formula_i` only for collecting same formulas images, the order doesn't matter.
a leading question: can we find images related to the same formula in different dataset segments?

**Comment by LachlanStuart** [comment]
@omerb01 If we look at the `huge4` dataset, the dataset segments are split like this:

segment # | lowest mz| highest mz
-- | -- | --
0 | 198.010676 | 200.958094
1 | 200.958094 | 201.091571
2 | 201.091571 | 202.094092
3 | 202.094092 | 202.990456
4 | 202.990456 | 203.087088
5 | 203.087088 | 203.980149
6 | 203.980149 | 204.434332
7 | 204.434332 | 204.989594
8 | 204.989594 | 205.083574
9 | 205.083574 | 205.880832
10 | 205.880832 | 206.001647
11 | 206.001647 | 206.982079

A typical formula will have 4 peaks, which are distributed something like this, when sorted by the mz of peak 0: 

`formula_i` | peak 0 `mz` | peak 1 `mz` | peak 2 `mz` | peak 3 `mz`
-- | -- | -- | -- | --
237106 | 200.006899 | 201.010312 | 202.004920 | 203.008290
145669 | 200.006987 | 201.006541 | 201.010394 | 202.003257
176945 | 200.007449 | 202.005614 | 204.004776 | 206.004942
620657 | 200.008169 | 202.003564 | 203.006799 | 204.007728


Or if you map them to segments numbers:

`formula_i` | peak 0 segment | peak 1 segment | peak 2segment | peak 3 segment
-- | -- | -- | -- | --
237106 | 0|1|2|4
145669 | 0|1|2|2
176945 | 0 | 2|6|11
620657 | 0 | 2|4|6

To evaluate all peaks for one formula, it's necessary to load up to 4 separate dataset segments (technically up to 8 because sometimes the mz +/- 3ppm range will sit on the border between two   segments). Although we could spend some time optimizing the order that formulas are iterated so that we get all formulas need e.g. `0,1,2,4` in one batch, then `0,2,6,11` in another batch, it would increase the I/O significantly, because segments would have to be unloaded before they've finished being used, e.g. with the above formulas:
* Load segments `0,1,2,4` to process the `0,1,2,4` and `0,1,2,2` batches of formulas
* Unload segment `4` and load segments `6,11` to process `0,2,6,11` batch
* Unload segment `11` and *wastefully reload* segment `4` to process the `0,2,4,6` batch

I feel a lot of time could be spent on trying to optimize that the processing order, but it would still be slow due to excess I/O.

There's also a third option - capture images in `mz` order and save them to COS in chunks based on `formula_i`, then calculate the metrics in a separate step that doesn't load the dataset segments at all. Although this would scale to much bigger datasets than either approach so far, it would probably also be slower due to the extra I/O.

**Comment by omerb01** [comment]
due to the complexity of reaching this target, I think we can rely on #54 and close this one.

**Pull request closed without merging**